Properly interpolate `#[openapi]`-annotation with `TokenStreams`

domenicquirl opened the pull request:
Replaces the String-based interpolation of the auto-discovered tokens into the manually annotated attributes with quoting via `TokenStream`s.

## What this PR fixes

When bumping to `utoipa` v5 / `utoipauto` v0.2, I noticed I was getting an error in my IDE (VS Code) on the `#[utoipauto]` annotation. Everything built fine when running tests / just building with `cargo`, but the macro crashed by `unwrap`ping an `Err` when run via rust-analyzer.

I was able to narrow the issue down to your handling of the original `#[openapi]` annotations on the struct deriving `OpenAPI`: `utoipauto` does its discovery and then pieces together the automatically discovered components with the existing annotations like `modifiers`, `tags`, etc. To do this, `update_openapi_macro_attributes` takes the whole attribute, converts it to a string via `to_token_stream` and then tries to strip `#[openapi(` at the start and `)]` at the end of the attribute, inserting everything in between into the newly generated annotation that is created as the output of the macro, containing also the automatically discovered paths. 

With `cargo`, this works because the `rustc` proc-macro server is relatively exact in its `to_string()` representation of the attribute: 

```rs
#[openapi(modifiers(&Addon1, &Addon2),\ninfo(title = \"Title\", version = \"v1\",),\nvariables((\"var\" =\n(description = \"Description\")),)),))]
```

(I've left out a bunch of stuff but you get the idea.)

Compare this to the `to_string()` representation we get when running under rust-analyzer:

```rs
# [openapi (modifiers (& Addon1 , & Addon2) , info (title = \"Title\" , version = \"v1\" ,) , variables ((\"var\" = (description = \"Description\")) ,)) ,))
```

As you can see, rust-analyzer's proc-macro server inserts additional whitespace in between the tokens inside that make up the token stream, which causes the substrings to no longer match and not be removed, and therefore causes the macro to output something like `#[openapi(paths(...), #[openapi(modifiers(...))] )]` containing first the auto-discovered paths, followed by _the entire original attribute_, which fails to re-parse as a `TokenStream` because attributes are not valid inside attributes.

Given that the [`impl Display for TokenStream`](https://doc.rust-lang.org/proc_macro/struct.TokenStream.html#impl-Display-for-TokenStream) explicitly cautions against matching against its output because it isn't stable by saying

> Note: the exact form of the output is subject to change, e.g. there might be changes in the whitespace used between tokens. Therefore, you should _not_ do any kind of simple substring matching on the output string (as produced by `to_string`) to implement a proc macro, because that matching might stop working if such changes happen. Instead, you should work at the `TokenTree` level, e.g. matching against `TokenTree::Ident`, `TokenTree::Punct`, or `TokenTree::Literal`.

I'm gonna suggest that this should be considered a bug in `utoipauto`.

## Changes made to fix this issue

The PR changes the interpolation inside `update_openapi_macro_attributes` to work on `TokenStream`s instead of `String`s. To do this, I had to create a larger change than I'd like, because I had to go back to `file_utils::exctract_module_name_from_path` to return an actual `Path` (this also fixes a bug where hyphens were only replaced with underscores in the crate name, but not module names, which for some reason existed in one of the tests), such that `discover_from_file` can now `parse_module_items` as `Path`s, such that `update_openapi_macro_attributes` can interpolate those paths into the original macro input on the level of `TokenStream`s / `quote`. This necessitated also rewriting the `extract_` and `remove_` functions in `attribute_utils` to work on `syn`'s `Attribute`/`Meta`.

I've ran the changes in the PR against your test suite, including the `acceptance` tests, and further verified the macro output is unchanged / correct in a regular build in the project I was having the issue in originally. I'll leave a comment for some test changes I had to make as a review.

domenicquirl commented on utoipauto-core/src/file_utils.rs:
```diff
@@ -110,95 +115,139 @@ fn find_segment_and_skip<'a>(segments: &'a [&str], to_find: &[&str], to_skip: us
+            extract_module_name_from_path("./utoipa-auto-macro/tests/controllers/controller1.rs", "crate")
+                .to_token_stream()
+                .to_string()
+                .replace(" ", ""),
```
The `replace` here is to adjust for additional whitespace between tokens when converting with `to_string`, which happened for me in the tests even when testing with `cargo` only.

domenicquirl commented on utoipauto-core/src/attribute_utils.rs:
```diff
@@ -27,257 +36,276 @@ pub fn update_openapi_macro_attributes(
-                &"".to_string(),
-                &"".to_string(),
+                Punctuated::new(),
+                &quote::quote!(crate::api::test),
```
I couldn't get `./src` to behave reasonably here (if you quote it as `"./src"`, it will be interpolated as a string literal, including quotation marks). I'm pretty sure after having worked through this that the path originally given wouldn't be what is passed here anyways, since it's only used for discovery and the result of discovery should be the resolved crate / module / item paths? But you might want to double-check that I didn't bork something here.

DenuxPlays replied:
Looks weird but fine to me.
At least I couldn't find anything broken in the tests or when I tested it locally so its probably fine.

domenicquirl commented:
(Had to fix a doc test I missed.)

DenuxPlays commented:
Hey I tested it.

It works great and actually fixes: https://github.com/ProbablyClem/utoipauto/issues/46

I just have to make a code review.
Hopefully I find time in like 2 hours or something like that.

Just as an update

DenuxPlays commented on utoipauto-core/src/attribute_utils.rs:
```diff
@@ -27,257 +36,276 @@ pub fn update_openapi_macro_attributes(
-    } else {
-        src_uto_macro
+fn remove_paths_and_components(nested_attributes: Punctuated<Meta, Token![,]>) -> TokenStream {
+    let mut remaining = Vec::with_capacity(nested_attributes.len());
```
```suggestion
    let mut remaining = Vec::new();
```
As we only add `responses` more memory then needed is being allocated.
We can take the small performance hit here.

domenicquirl replied:
You're right, I think I changed this before changing the function body and then it stuck. Reverted it back to just `Vec::new` with the last commit.

DenuxPlays reviewed (changes requested):
Just one comment but it looks great.

But in the extract_* functions I am not sure if we should use nested for loops or if we should work with iterators.
Ofcourse both work so it's just a preferene.

But I had something like this in my mind:
```rust
fn extract_schemas(nested_attributes: &Punctuated<Meta, Token![,]>) -> TokenStream {
    nested_attributes
        .iter()
        .find_map(|meta| {
            let Meta::List(list) = meta else { return None };
            if !list.path.is_ident("components") {
                return None;
            }
            
            let nested = list
                .parse_args_with(Punctuated::<Meta, Token![,]>::parse_terminated)
                .expect("Expected a list of attributes inside components(...)!");

            nested.iter().find_map(|meta| {
                let Meta::List(list) = meta else { return None };
                if list.path.is_ident("schemas") {
                    Some(list.tokens.clone())
                } else {
                    None
                }
            })
        })
        .unwrap_or_else(TokenStream::new)
}
```

Maybe @ProbablyClem can say what he wants as I don't really care.
Both work and are readable.

domenicquirl commented:
Addressed your comments and just went ahead with changing the `extract_` methods to iterators. It's a bit longer but less indented, so that's nice at least. I also pulled together schemas and responses into a single function as they only differ in the final key / ident that they are looking for.

Happy to hear that this actually fixes the other issue as well ☺️

DenuxPlays commented:
LGTM

Thank you for your contribution.

FYI:
I would like to put this pr through a test phase together with: https://github.com/ProbablyClem/utoipauto/pull/47
So this will probably be in the release `0.3.0-alpha.1`
